compiler/ast/ast.go

```diff
@@ -398,7 +397,7 @@ func (o *OpDecl) End() int { return o.Rparen }
 type TypeDecl struct {
 	Kind       string      `json:"kind" unpack:""`
 	KeywordPos int         `json:"keyword_pos"`
-	Name       string      `json:"name"`
+	Name       Pattern     `json:"name"`
```

Comment:
`Pattern` feels very wrong since this can't be a `Glob` or a `Regexp`. Are you using it because `String` and `QuotedString` both implement it?

Reply:
I agree it is really weird. Yes I'm using it because String and QuotedString implement it. Open to suggestions.

Reply:
I mean ideally this would just be an identifier but we are in need of a stop gap until that decision can be made.

Reply:
(The `TypeDecl` rule in parser.peg could then go something like this.

```
TypeDecl
  = "type" _ id:(Identifier / QuotedIdentifier) __ "=" __ typ:Type {
      ...
    }

QuotedIdentifier
  = name:QuotedString {
      return &ast.ID{
          Kind: "ID",
          Name: name.(string),
          NamePos: c.pos.offset,
      }, nil
    }
```
)

Reply:
The problem with this approach is you now have inaccurate node positioning for `QuotedIdentifier` given that it won't take into account the space of the quotes when finding the location of end.

Reply:
I suppose the least lazy, most correct way would be to create a new interface that supports both ID and QuotedString. It's not great but neither is this feature IMHO.

Reply:
> The problem with this approach is you now have inaccurate node positioning for QuotedIdentifier given that it won't take into account the space of the quotes when finding the location of end.

Could we say `NamePos: c.pos.offset+1` (instead of `c.pos.offset`) and get errors that look like this (where the tildes don't extend beneath the quotation marks)?

```
type "something bad" = int64
      ~~~~~~~~~~~~~
```

Seems fine to me if so.

Reply:
I still hate it, but ok.

Reply:
As long as we both can agree that this is a hack and not a rock solid solution, I'm ok with this.

Reply:
We can agree.

And you're right that a new interface supporting both ID and QuotedString is the most correct way to solve this. I just don't think it's worth its weight.